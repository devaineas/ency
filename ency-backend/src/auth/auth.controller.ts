import {
  Body,
  Controller,
  Get,
  InternalServerErrorException,
  NotFoundException,
  Post,
  UseInterceptors,
} from "@nestjs/common";
import { User } from "@prisma/client";
import { PrismaInterceptor } from "src/prisma/prisma.interceptor";
import { AuthService } from "./auth.service";
import { FirebaseUser } from "./decorators/user.decorator";
import { SignUpDto } from "./dto/sign-up.dto";
import { IFirebaseUser } from "./interfaces/user.interface";

@Controller("auth")
@UseInterceptors(PrismaInterceptor)
export class AuthController {
  constructor(private authService: AuthService) {}

  @Post("sign-up")
  async signUp(
    @FirebaseUser() user: IFirebaseUser,
    @Body() body: SignUpDto,
  ): Promise<User> {
    const { email, name } = body;

    return await this.authService.createUser(user.uid, name, email);
  }

  @Get()
  async getUser(@FirebaseUser() user: IFirebaseUser): Promise<User> {
    const userData = await this.authService.getUser(user.uid);

    if (!userData) throw new NotFoundException("User with given id not found");
    return userData;
  }
}
