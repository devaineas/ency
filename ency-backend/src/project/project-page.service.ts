import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { ProjectPage } from "@prisma/client";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import { PrismaService } from "src/prisma/prisma.service";
import { CreateProjectPageDto } from "./dto/create-project-page.dto";
import { UpdateProjectPageDto } from "./dto/update-project-page.dto";
import {
  ProjectPageEntity,
  ProjectPageEntityNoData,
} from "./enitiies/projectPage.entity";

@Injectable()
export class ProjectPageService {
  constructor(private prisma: PrismaService) {}

  create(
    user: IFirebaseUser,
    createProjectPageDto: CreateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    return this.prisma.projectPage.create({
      data: {
        name: "Untitled Page",
        data: "",
        project: {
          connect: {
            // Only the owner can add
            id_userUid: {
              id: createProjectPageDto.projectId,
              userUid: user.uid,
            },
          },
        },
      },
    });
  }

  findAll(
    user: IFirebaseUser,
    projectId: number,
  ): Promise<ProjectPageEntityNoData[]> {
    return this.prisma.projectPage.findMany({
      where: {
        project: {
          id: projectId,
          OR: [
            // Either thr project isPublic or is owned by the user
            {
              isPublic: true,
            },
            {
              userUid: user.uid,
            },
          ],
        },
      },
      select: {
        // Deselecting data
        id: true,
        name: true,
        projectId: true,
        data: false,
      },
    });
  }

  async findOne(
    user: IFirebaseUser,
    pageId: number,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findFirst({
      where: {
        id: pageId,
        OR: [
          // Either project isPublic or owned by the authorized user
          {
            project: {
              isPublic: true,
            },
          },
          {
            project: {
              userUid: user.uid,
            },
          },
        ],
      },
    });

    if (!projectPage) {
      throw new NotFoundException("Project with given id not found");
    }

    return projectPage;
  }

  async update(
    user: IFirebaseUser,
    pageId: number,
    updateProjectPageDto: UpdateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findUnique({
      where: {
        id: pageId,
      },
      select: {
        project: true,
      },
    });

    if (!projectPage)
      throw new NotFoundException("Page with given id does not exist");

    if (projectPage?.project.userUid === user.uid)
      throw new BadRequestException(
        "You are not authorized to perform this action",
      );

    return this.prisma.projectPage.update({
      where: {
        id: pageId,
      },
      data: {
        ...updateProjectPageDto,
      },
    });
  }

  async remove(
    user: IFirebaseUser,
    pageId: number,
  ): Promise<ProjectPageEntity> {
    const projectPage = await this.prisma.projectPage.findUnique({
      where: {
        id: pageId,
      },
      select: {
        project: true,
      },
    });

    if (!projectPage)
      throw new NotFoundException("Page with given id does not exist");

    if (projectPage?.project.userUid === user.uid)
      throw new BadRequestException(
        "You are not authorized to perform this action",
      );

    return this.prisma.projectPage.delete({
      where: {
        id: pageId,
      },
    });
  }
}
