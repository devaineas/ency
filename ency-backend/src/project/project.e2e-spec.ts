import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";
import Request from "supertest";
import { ProjectModule } from "./project.module";

describe("ProjectController (e2e)", () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [ProjectModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it("/ (GET)", () => {
    return (
      Request(app.getHttpServer())
        // TODO
        .get("/")
        .expect(404)
    );
  });
});
