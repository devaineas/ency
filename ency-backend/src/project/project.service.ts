import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import { PrismaService } from "../prisma/prisma.service";
import { CreateProjectDto } from "./dto/create-project.dto";
import { UpdateProjectDto } from "./dto/update-project.dto";
import { ProjectEntity } from "./enitiies/project.entity";

const FIRST_PAGE_TITLE = "Untitled page 1";

@Injectable()
export class ProjectService {
  constructor(private prisma: PrismaService) {}

  async create(
    createProjectDto: CreateProjectDto,
    userUid?: string,
  ): Promise<ProjectEntity> {
    if (!userUid) throw new BadRequestException();

    const project = await this.prisma.project.create({
      data: {
        ...createProjectDto,
        user: {
          connect: {
            uid: userUid,
          },
        },
        pages: {
          create: {
            name: FIRST_PAGE_TITLE,
            data: "",
          },
        },
      },
    });

    return {
      ...project,
      pageCount: 1,
    };
  }

  async findAll(
    user: IFirebaseUser,
    userUid: string,
  ): Promise<ProjectEntity[]> {
    const projects = await this.prisma.project.findMany({
      where: {
        userUid,
        OR: [
          // Either Project isPublic or is owned by authorized user
          {
            isPublic: true,
          },
          {
            userUid: user.uid,
          },
        ],
      },
    });

    // ISSUE: Change this to more performant one once this issue is resolved: https://github.com/prisma/prisma/issues/5079
    return Promise.all(
      projects.map(async (project) => {
        const pageCount = await this.findOnePageCount(project.id);
        return {
          ...project,
          pageCount,
        };
      }),
    );
  }

  async findOne(id: number): Promise<ProjectEntity> {
    const project = await this.prisma.project.findUnique({
      where: {
        id,
      },
    });

    if (!project)
      throw new NotFoundException(
        `A Project with the given id: ${id} not found`,
      );

    const pageCount = await this.findOnePageCount(project.id);

    return {
      ...project,
      pageCount,
    };
  }

  async findOnePageCount(projectId: number): Promise<number> {
    return this.prisma.projectPage.count({
      where: {
        projectId: projectId,
      },
    });
  }

  async update(
    userUid: string,
    id: number,
    updateProjectDto: UpdateProjectDto,
  ): Promise<ProjectEntity> {
    return await this.prisma.project.update({
      where: {
        id_userUid: {
          id: id,
          userUid: userUid,
        },
      },
      data: {
        ...updateProjectDto,
      },
    });
  }

  async remove(userUid: string, id: number): Promise<number> {
    // return this.prisma.project.delete({
    //   where: {
    //     id_userUid: {
    //       id: id,
    //       userUid: userUid,
    //     },
    //   },
    // });

    const res = await this.prisma.$executeRaw`
      DELETE FROM project WHERE "userUid" = ${userUid} AND id = ${id}
    `;

    if (res !== 1) {
      throw new NotFoundException("Project with given id not found");
    }

    return res;
  }
}
