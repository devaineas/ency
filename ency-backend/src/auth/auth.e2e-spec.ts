import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import Request from 'supertest';
import { AuthModule } from './auth.module';

describe('AuthController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AuthModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });
// TODO
//   it('/ (GET)', () => {
//     return Request(app.getHttpServer())
//       .get('/')
//       .expect(404)
//   });
});
