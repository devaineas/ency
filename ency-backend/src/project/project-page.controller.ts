import {
  Controller,
  Get,
  Post,
  Body,
  Put,
  Param,
  Delete,
  ParseIntPipe,
} from "@nestjs/common";
import { ProjectPageService } from "./project-page.service";
import { CreateProjectPageDto } from "./dto/create-project-page.dto";
import { UpdateProjectPageDto } from "./dto/update-project-page.dto";
import { Public } from "src/auth/auth.guard";
import { FirebaseUser } from "src/auth/decorators/user.decorator";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import {
  ProjectPageEntity,
  ProjectPageEntityNoData,
} from "./enitiies/projectPage.entity";

/**
 * Here id param will be used to refer to the Projects ID and pageId for ProjectPages ID
 */
@Controller("project")
export class ProjectPageController {
  constructor(private readonly projectPageService: ProjectPageService) {}

  @Post(":id/page")
  create(
    @FirebaseUser() user: IFirebaseUser,
    @Body() createProjectPageDto: CreateProjectPageDto,
  ) {
    return this.projectPageService.create(user, createProjectPageDto);
  }

  @Get(":id/page")
  @Public()
  findAll(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) projectId: number,
  ): Promise<ProjectPageEntityNoData[]> {
    return this.projectPageService.findAll(user, projectId);
  }

  @Get(":id/page/:pageId")
  @Public()
  findOne(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.findOne(user, pageId);
  }

  @Put(":id/page/:pageId")
  update(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
    @Body() updateProjectPageDto: UpdateProjectPageDto,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.update(user, pageId, updateProjectPageDto);
  }

  @Delete(":id/page/:pageId")
  remove(
    @FirebaseUser() user: IFirebaseUser,
    @Param("pageId", ParseIntPipe) pageId: number,
  ): Promise<ProjectPageEntity> {
    return this.projectPageService.remove(user, pageId);
  }
}
