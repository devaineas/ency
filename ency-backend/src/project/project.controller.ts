import {
  Controller,
  Get,
  Post,
  Body,
  Put,
  Param,
  Delete,
  ParseIntPipe,
  NotFoundException,
} from "@nestjs/common";
import { ProjectService } from "./project.service";
import { CreateProjectDto } from "./dto/create-project.dto";
import { UpdateProjectDto } from "./dto/update-project.dto";
import { FirebaseUser } from "../auth/decorators/user.decorator";
import { IFirebaseUser } from "../auth/interfaces/user.interface";
import { Public } from "../auth/auth.guard";
import { ProjectEntity } from "./enitiies/project.entity";

@Controller("projects")
export class ProjectController {
  constructor(private readonly projectService: ProjectService) {}

  @Post()
  create(
    @Body() createProjectDto: CreateProjectDto,
    @FirebaseUser() user?: IFirebaseUser,
  ): Promise<ProjectEntity> {
    return this.projectService.create(createProjectDto, user?.uid);
  }

  @Get()
  findAll(@FirebaseUser() user: IFirebaseUser): Promise<ProjectEntity[]> {
    return this.projectService.findAll(user, user.uid);
  }

  @Public()
  @Get("user/:uid")
  async findUsersProjects(
    @FirebaseUser() user: IFirebaseUser,
    @Param("uid") userUid: string,
  ): Promise<ProjectEntity[]> {
    return this.projectService.findAll(user, userUid);
  }

  @Public()
  @Get(":id")
  async findOne(
    @Param("id", ParseIntPipe) id: number,
    @FirebaseUser() user?: IFirebaseUser,
  ): Promise<ProjectEntity> {
    const project = await this.projectService.findOne(id);

    if (!project.isPublic && project.userUid !== user?.uid) {
      throw new NotFoundException(
        "A project with the given id not found (or is not public and not owned by you)",
      );
    }

    return project;
  }

  @Put(":id")
  update(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) id: number,
    @Body() updateProjectDto: UpdateProjectDto,
  ): Promise<ProjectEntity> {
    return this.projectService.update(user.uid, id, updateProjectDto);
  }

  @Delete(":id")
  remove(
    @FirebaseUser() user: IFirebaseUser,
    @Param("id", ParseIntPipe) id: number,
  ) {
    return this.projectService.remove(user.uid, id);
  }
}
