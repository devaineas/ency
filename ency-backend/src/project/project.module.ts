import { Module } from "@nestjs/common";
import { ProjectService } from "./project.service";
import { ProjectController } from "./project.controller";
import { PrismaModule } from "../prisma/prisma.module";
import { AuthModule } from "../auth/auth.module";
import { ProjectPageController } from "./project-page.controller";
import { ProjectPageService } from "./project-page.service";

@Module({
  imports: [PrismaModule, AuthModule],
  controllers: [ProjectController, ProjectPageController],
  providers: [ProjectService, ProjectPageService],
})
export class ProjectModule {}
