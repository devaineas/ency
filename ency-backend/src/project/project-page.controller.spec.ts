import { Test, TestingModule } from "@nestjs/testing";
import { AppModule } from "src/app.module";
import { AuthModule } from "src/auth/auth.module";
import { PrismaModule } from "src/prisma/prisma.module";
import { ProjectPageController } from "./project-page.controller";
import { ProjectPageService } from "./project-page.service";
import { ProjectController } from "./project.controller";
import { users, projects } from "../../prisma/mock-data.json";
import { IFirebaseUser } from "src/auth/interfaces/user.interface";
import { ProjectService } from "./project.service";

describe("ProjectPageController", () => {
  let projectController: ProjectController;
  let controller: ProjectPageController;
  let module: TestingModule;

  let newProjectId: number;
  const mockUserData = users[0];
  const user: IFirebaseUser = { uid: mockUserData.uid };

  beforeAll(async () => {
    module = await Test.createTestingModule({
      imports: [PrismaModule, AuthModule, AppModule],
      controllers: [ProjectPageController, ProjectController],
      providers: [ProjectPageService, ProjectService],
    }).compile();

    controller = module.get<ProjectPageController>(ProjectPageController);
    projectController = module.get<ProjectController>(ProjectController);

    const mockProject = projects.filter((project) => project.isPublic)[0];
    const { userUid, ...mockProjectData } = mockProject;
    const newProjectData = await projectController.create(
      mockProjectData,
      user,
    );
    newProjectId = newProjectData.id;
  });

  afterAll(async () => {
    await module.close();
  });

  describe("Post(:id/page) create", () => {
    it("With valid input, Should return valid Response", async () => {
      const newProjectPage = await controller.create(user, {
        projectId: newProjectId,
      });

      expect(newProjectPage).toMatchSnapshot("new-project-page");
    });
  });
});
