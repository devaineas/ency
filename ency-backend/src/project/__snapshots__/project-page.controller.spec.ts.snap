// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPageController Post(:id/page) create With valid input, Should return valid Response: new-project-page 1`] = `
Object {
  "data": "",
  "id": 2,
  "name": "Untitled Page",
  "projectId": 1,
}
`;
